docs/rules/replace-dynamic-query-by-prepared-statement.md

```diff
@@ -0,0 +1,108 @@
+links:
+    - displayName: "Prepared Statements (with Parameterized Queries)"
+      url: "https://cheatsheetseries.owasp.org/cheatsheets/SQL_Injection_Prevention_Cheat_Sheet.html#defense-option-1-prepared-statements-with-parameterized-queries"
+    
+description:
+    Rule description.
+tags: ["Security"]
+---
+
+# Replace Dynamic Query by PreparedStatement
+
+[[toc]]
+
+## Properties
+
+<RuleProperties />
+
+## Description
+
+Replaces [java.sql.Statement](https://docs.oracle.com/javase/8/docs/api/java/sql/Statement.html) with [java.sql.PreparedStatement](https://docs.oracle.com/javase/8/docs/api/java/sql/PreparedStatement.html) (aka parameterized queries) if the SQL query is constructed by concatenating string literals with user defined expressions (e.g. variables, method invocations, user input, etc). 
+Parameterized queries enforce a distinction between the SQL code and the data passed through parameters. 
+
+## Benefits
+
+Prevents SQL injections.
+
+## Requirement & Tags
+
+::: warning Requirements
+Java 1.1
+:::
+
+::: tip Tags
+<TagLinks />
+:::
+
+## Code Changes
+
+
+### Using Statement execute
+
+__Pre__
+```java
+String departmentId = "40 OR '1'='1";
+String query = "SELECT first_name FROM employee WHERE department_id ='" + departmentId + "' ORDER BY last_name";
+Statement statement = connection.createStatement();
+statement.execute(query);
+ResultSet resultSet = statement.getResultSet();
+```
+
+__Post__
+```java
+String departmentId = "40 OR '1'='1";
+String query = "SELECT first_name FROM employee WHERE department_id = ?" + " ORDER BY last_name";
+PreparedStatement statement = connection.prepareStatement(query);
+statement.setString(1, departmentId);
+ResultSet resultSet = statement.executeQuery();
+```
+
+### Using Statement executeQuery
+
+__Pre__
+```java
+String departmentId = "40 OR '1'='1";
+String query = "SELECT first_name FROM employee WHERE department_id ='" + departmentId + "' ORDER BY last_name";
+Statement statement = connection.createStatement();
+ResultSet resultSet = statement.executeQuery(query);
+```
+
+__Post__
+```java
+String departmentId = "40 OR '1'='1";
+String query = "SELECT first_name FROM employee WHERE department_id = ?" + " ORDER BY last_name";
+PreparedStatement statement = connection.prepareStatement(query);
+statement.setString(1, departmentId);
+ResultSet resultSet = statement.executeQuery();
+```
+
+### Multiple Concatenation Lines
+
+__Pre__
+```java
+String departmentId = "40 OR '1'='1";
+int id = 10;
+String query = "SELECT first_name FROM employee WHERE  id > '" + id + "'";
+query += " AND department_id ='" + departmentId + "' ORDER BY last_name";
+Statement statement = connection.createStatement();
+ResultSet resultSet = statement.executeQuery(query);
+```
+
+__Post__
+```java
+String departmentId = "40 OR '1'='1";
+int id = 10;
+String query = "SELECT first_name FROM employee WHERE  id >  ?" + "";
+query += " AND department_id ='" + departmentId + "' ORDER BY last_name";
```

Comment:
Interesting idea: maybe it is possible to replace also the second parameter after the '+=' concatenation in 
```java
query += " AND department_id ='" + departmentId + "' ORDER BY last_name";
```
for example by recognizing that the statement beginning with  'query +=' immediately follows the initialization statement of the query.

Reply:
yes yes. I have that in mind. I will fix that. Thanks for pointing it out again.